Tests/MongoSwiftSyncTests/UnifiedTestRunner/Matching.swift

```diff
@@ -361,6 +363,226 @@ func matchesEvents(expected: [ExpectedEvent], actual: [CommandEvent], entities:
+            try context.withPushedElt("isClientError") {
+                if isClientError {
+                    guard actualIsClientError else {
+                        throw NonMatchingError(
+                            expected: "Error to be a client error",
+                            actual: self,
+                            context: context
+                        )
+                    }
+                } else {
+                    guard !actualIsClientError else {
+                        throw NonMatchingError(
+                            expected: "Error to not be a client error",
+                            actual: self,
+                            context: context
+                        )
+                    }
+                }
+            }
+        }
+
+        if let errorContains = expected.errorContains {
+            try context.withPushedElt("errorContains") {
+                guard let mongoError = self as? MongoErrorProtocol else {
+                    throw NonMatchingError(
+                        expected: "error to conform to MongoErrorProtocol",
+                        actual: self,
+                        context: context
+                    )
+                }
+                guard mongoError.errorDescription!.lowercased().contains(errorContains.lowercased()) else {
+                    throw NonMatchingError(
+                        expected: "error message to contain \(errorContains)",
+                        actual: mongoError,
+                        context: context
+                    )
+                }
+            }
+        }
+
+        if let errorCode = expected.errorCode {
+            try context.withPushedElt("errorCode") {
+                guard let actualWithCodes = self as? HasErrorCodes else {
+                    throw NonMatchingError(
+                        expected: "error to have error code(s)",
+                        actual: self,
+                        context: context
+                    )
+                }
+
+                guard actualWithCodes.errorCodes.contains(errorCode) else {
+                    throw NonMatchingError(
+                        expected: "error to have error code \(errorCode)",
+                        actual: actualWithCodes,
+                        context: context
+                    )
+                }
+            }
+        }
+
+        if let codeName = expected.errorCodeName {
+            try context.withPushedElt("errorCodeName") {
+                guard let actualWithCodeNames = self as? HasErrorCodes else {
+                    throw NonMatchingError(
+                        expected: "error to have error code(s)",
+                        actual: self,
+                        context: context
+                    )
+                }
+                // TODO: SWIFT-1022: Due to CDRIVER-3147 many of our errors are currently missing code names, so we
```

Comment:
this is kind of not great in the case of a bulk error since you could mistakenly take the presence of an empty string that is really present because of some different error code/name pair to mean we had the right code name here. that said, I'm not too worried about it because we do make the error code number assertions correctly.

CDRIVER-3147 has also been slated to be fixed in 1.18 so I don't think we'll have this issue for very long.